types/validation.go

```diff
@@ -273,11 +278,16 @@ func verifyCommitBatch(
 		// Validate signature.
 		voteSignBytes := commit.VoteSignBytes(chainID, int32(idx))
 
-		// add the key, sig and message to the verifier
-		if err := bv.Add(val.PubKey, voteSignBytes, commitSig.Signature); err != nil {
-			return err
+		cacheKey := string(val.Address) + string(voteSignBytes)
+		_, ok := verifiedSignatureCache[cacheKey]
```

Comment:
Nice! This has a malleabiltiy vector though if addresses weren't fixed size. Its also not zero-allocation, which is good to have at the sig verification level.

Perhaps structure the map using vote sign bytes as the key, and then the val address as the value? Not sure we can avoid the string casts for immutability reasons, so maybe we just live with the allocs. Should see how much it slows it down in the normal case.

Also can we allow passing in a nil map (skipping the check if nil)

Reply:
Updated to make these changes.
- Using `string(voteSignBytes)` as the key and val address as the value
- Check `verifiedSignatureCache != nil` before any usage. Pass `nil` where ever the caching does not need to be used